src/cluster.c

```diff
@@ -852,23 +692,231 @@ static dict *parse_cluster_slots(valkeyClusterContext *cc, valkeyReply *reply) {
     return NULL;
 }
 
+/* Keep lists of parsed replica nodes in a dict using the primary_id as key. */
+static int retain_replica_node(dict *replicas, char *primary_id, valkeyClusterNode *node) {
+    sds key = sdsnew(primary_id);
+    if (key == NULL)
+        return VALKEY_ERR;
+
+    struct hilist *replicaList;
+
+    dictEntry *de = dictFind(replicas, key);
+    if (de == NULL) {
+        /* Create list to hold replicas for a primary. */
+        replicaList = listCreate();
+        if (replicaList == NULL) {
+            sdsfree(key);
+            return VALKEY_ERR;
+        }
+        replicaList->free = listClusterNodeDestructor;
+        if (dictAdd(replicas, key, replicaList) != DICT_OK) {
+            sdsfree(key);
+            listRelease(replicaList);
+            return VALKEY_ERR;
+        }
+    } else {
+        sdsfree(key);
+        replicaList = dictGetEntryVal(de);
+    }
+
+    if (listAddNodeTail(replicaList, node) == NULL)
+        return VALKEY_ERR;
+
+    return VALKEY_OK;
+}
+
+/* Store parsed replica nodes in the primary nodes, which holds a list of replica
+ * nodes. The `replicas` dict shall contain lists of nodes with primary_id as key. */
+static int store_replica_nodes(dict *nodes, dict *replicas) {
+    if (replicas == NULL)
+        return VALKEY_OK;
+
+    dictIterator di;
+    dictInitIterator(&di, nodes);
+    dictEntry *de;
+    while ((de = dictNext(&di))) {
+        valkeyClusterNode *primary = dictGetEntryVal(de);
+
+        /* Move replica nodes related to this primary. */
+        dictEntry *der = dictFind(replicas, primary->name);
+        if (der != NULL) {
+            assert(primary->slaves == NULL);
+            /* Move replica list from replicas dict to nodes dict. */
+            primary->slaves = dictGetEntryVal(der);
+            dictSetHashVal(replicas, der, NULL);
```

Comment:
I don't recognize the function name `dictSetHashVal`. It must be copied from a very old Redis version. :smile: